### PR TITLE
Add TypeScript Loader

### PR DESCRIPTION
**Pull request by julian-londono**
This PR adds the first iteration of the TypeScript loader to `presm`.  This loader allows on-the-fly transpilation and execution of TypeScript files.  This loader should resolve imported specifiers using TypeScript logic (for `.ts`, `.js` and extensionless imports).

This PR also begins adding a comprehensive and living set of tests in `examples/tests/tests.ts`.  These can be run using:
`npm run test`.  

NOTE: These tests and the infrastructure around them are a WIP; 2 tests regarding dynamic importing are currently failing

### CODE REVIEW

File: examples/tests/tests.ts [hunk 1]
```diff
@@ -4,11 +4,29 @@ const loadViaRequire = createRequire(import.meta.url);
 // import test from 'tape';
```
Comment (jkrems) [inline examples/tests/tests.ts:4]: I think I would usually expect the tests to be either in top-level `test/` directory or in test files next to the files under test (`src/my_actual_file_test.ts`).

File: examples/loaders/preprocessor-typescript.mjs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            }
+          }
+
+          return node => ts.visitNode(node, visit);
+        };
+      }
+
+      // Add resolver to hook into transpiler logic "before"
+      // transpiling occurs
+      configOptions.transformers = {before: [tsModuleResolver()]};
```
Comment (jkrems) [inline examples/loaders/preprocessor-typescript.mjs:59]: Style nit: I generally prefer copy over mutate. Especially for arguments it can be problematic to mutate the value - the caller of this function may not expect the object to change.

Alternative:

```js
configOptions = {
  ...configOptions,
  transformers: {before: [tsModuleResolver()]},
};
```

File: examples/loaders/preprocessor-typescript.mjs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+      // Transpile source
+      source = ts.transpileModule(source, configOptions).outputText;
+
+      return {
+        source: source,
+        extension: outputExtensionTypes[0], // Example of choosing
+        // a valid extension from possible extensions}
```
Comment (jkrems) [inline examples/loaders/preprocessor-typescript.mjs:67]: If the length of the comment causes a line-break, it may be better to put the comment on the line before. Also - trailing `}` by accident?

```suggestion
        // Example of choosing a valid extension from possible extensions
        extension: outputExtensionTypes[0],
```

File: examples/calc/calc_runner.ts [hunk 4]
```diff
@@ -11,6 +11,6 @@ const { argv } = loadViaRequire('yargs')
   .alias('d', 'div')
   .array('d');
 
-import { calc } from './calc.ts';
+import { calc } from './calc.js';
 
-console.log(calc(argv));
+console.log(calc(argv));
```
Comment (jkrems) [inline examples/calc/calc_runner.ts:16]: Newline?

File: tsconfig.json [hunk 5]
```diff
@@ -3,9 +3,5 @@
   "compilerOptions": {
     "rootDir": ".",
     "outDir": "build"
-  },
```
Comment (jkrems) [inline tsconfig.json:6]: Do you remember why this change was necessary?

File: examples/calc/ops.ts [hunk 6]
```diff
@@ -14,15 +14,11 @@ function div(a: number, b: number): number {
   return a / b;
 }
 
-export const ops_list: { [x: string]: (a: number, b: number) => number } = {
+export const ops_list: {[x: string]: (a: number, b: number) => number} = {
```
Comment (jkrems) [inline examples/calc/ops.ts:17]: nit: `opsList`? Similar below, camelCase is more typical than snake_case in JS/TS.

File: examples/calc/calc.ts [hunk 7]
```diff
@@ -1,6 +1,6 @@
-import { ops_list, ops_symbols_list, argvType } from './ops.ts';
+import {ops_list, ops_symbols_list} from './ops';
 
-export function calc(argv: argvType): string {
+export function calc(argv: {[x: string]: string[]}): string {
```
Comment (feywind) [inline examples/calc/calc.ts:3]: I guess the comment I had from this morning was just that it's nice to break complex types like this out into a `type Foo =` declaration, or an `interface`. Then you can also name it to give others an idea of what you had in mind with it.

File: examples/calc/ops.ts (hunk 6 above)
Comment (feywind) [inline examples/calc/ops.ts:17]: Same comment here with the complex type.

File: examples/loaders/preprocessor-typescript.mjs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+export let sourceExtensionTypes = ['.ts', '.tsx'];
+
+export let outputExtensionTypes = ['.mjs', '.cjs'];
+
+export function getPreProcessor(configOptions = {}) {
+  return {
+    process(source, url) {
+      // Return new, resolved specifier using ts-specific
+      // logic if one exists; undefined otherwise
+      function getNewSpecficier(oldSpecifier) {
```
Comment (feywind) [inline examples/loaders/preprocessor-typescript.mjs:12]: Typo :)

File: examples/loaders/preprocessor-typescript.mjs (hunk 2 above)
Comment (julian-londono) [reply examples/loaders/preprocessor-typescript.mjs]: Good point, updated this and accounted for caller having their own transformers in `before` as well

File: examples/calc/calc_runner.ts (hunk 4 above)
Comment (julian-londono) [reply examples/calc/calc_runner.ts]: Cleaned up

File: tsconfig.json (hunk 5 above)
Comment (julian-londono) [reply tsconfig.json:6]: Not exactly, leaving this file as is for now as it doesn't currently affect anything

File: examples/calc/ops.ts (hunk 6 above)
Comment (julian-londono) [reply examples/calc/ops.ts]: Good tip, updated var names everywhere now

File: examples/calc/calc.ts (hunk 7 above)
Comment (julian-londono) [reply examples/calc/calc.ts:3]: Good call, I actually had an `interface` for this defined in `.ops.ts` but was getting an error because after transpiling the interface didn't exist.  This is a bug I will address but for now I'll move that interface to `calc.ts` itself.

File: examples/calc/ops.ts (hunk 6 above)
Comment (julian-londono) [reply examples/calc/ops.ts]: Same goes here, I'll define an `interface`

File: examples/loaders/preprocessor-typescript.mjs (hunk 8 above)
Comment (julian-londono) [reply examples/loaders/preprocessor-typescript.mjs]: Good catch, fixed :)

File: examples/tests/tests.ts (hunk 1 above)
Comment (julian-londono) [reply examples/tests/tests.ts:4]: I see, I'll address this along with a more comprehensive set of unit tests in a new PR today

File: examples/calc/calc.ts [hunk 9]
```diff
@@ -1,6 +1,11 @@
 import {opsList, opsSymbolsList} from './ops';
 
-export function calc(argv: {[x: string]: string[]}): string {
+ 
+interface argvType {
```
Comment (jkrems) [inline examples/calc/calc.ts:4]: nit: You'd usually UpperCase the names of types.

```suggestion
interface ArgvType {
```

Comment (jkrems) [reply examples/calc/calc.ts:4]: (And then you can also call it `Argv` because its' no longer easily confused with a value of that type. :))

Comment (feywind) [reply examples/calc/calc.ts:4]: Thanks! I knew I'd seen that somewhere and then couldn't find it.

**Comment by julian-londono** [comment]
Although this may not be the final version of my example `TypeScript` loader for PRESM, I will merge this PR and continue to iterate from there.

### Summary of PRs coming along with this PR:
- #8 Adds initial unit tests to PRESM components
- #10 Converts from `tape` to [`tap`](https://node-tap.org/) for all testing
  - adds system end-to-end tests for the loader
  - adds ability to mock the `loaderconfig.json` file (that controls user behavior)
  - re-architects loader structure using industry best practices as discussed with my hosts

**Comment by julian-londono** [comment]
All tests currently passing, one test (one TS unit test) disabled
